src/IOHprofiler_problem.hpp

```diff
@@ -49,40 +50,93 @@ template <class InputType> class IOHprofiler_problem {
+  int best_so_far_transformed_evaluations; 
 
-  /*still needs to think how to settle the type of this. */
-  //IOHprofiler_logger logger = NULL;
+  IOHprofiler_transformation transformation;
+
+  //std::size_t number_of_constraints;
+  //std::vector<double> constraints;
 
+  // A common function for evaluating fitness of problems.
+  // Raw evaluate process, tranformation operations, and logging process are excuted 
+  // in this function.
   void evaluate(std::vector<InputType> x, std::vector<double> &y) {
-    if(instance_id >= 1 && instance_id < 50) { 
-      transform_vars_xor(x,instance_id);
-    } else if(instance_id >= 50 && instance_id < 100) {
-      transform_vars_sigma(x,instance_id);
+    if(evaluations == 0) {
+      best_so_far_raw_objectives.reserve(number_of_objectives);
+      best_so_far_transformed_objectives.reserve(number_of_objectives);
+      for(int i = 0; i !=  number_of_objectives; ++i) {
+        best_so_far_raw_objectives.push_back(DBL_MIN_EXP);
+        best_so_far_transformed_objectives.push_back(DBL_MIN_EXP);
+      }
+
+      compareVector(best_variables,best_transformed_variables);
+      if(instance_id > 1 && instance_id <= 50) { 
+        transformation.transform_vars_xor(best_transformed_variables,instance_id);
+      } else if(instance_id > 50 && instance_id <= 100) {
+        transformation.transform_vars_sigma(best_transformed_variables,instance_id);
+      }
+      internal_evaluate(best_transformed_variables,transformed_optimal);
+      if(instance_id > 1) {
+        transformation.transform_obj_scale(transformed_optimal,instance_id);
+        transformation.transform_obj_shift(transformed_optimal,instance_id);
+      }
+
+    }
+    ++evaluations;
+
+    if(instance_id > 1 && instance_id <= 50) { 
+      transformation.transform_vars_xor(x,instance_id);
+    } else if(instance_id > 50 && instance_id <= 100) {
+      transformation.transform_vars_sigma(x,instance_id);
     }
     
     internal_evaluate(x,y);
     
+    raw_objectives.reserve(number_of_objectives);
     copyVector(y,raw_objectives);
-    transform_obj_scale(y,instance_id);
-    transform_obj_shift(y,instance_id);
+    if(compareObjectives(y,best_so_far_raw_objectives)) {
+      copyVector(y,best_so_far_raw_objectives);
+      best_so_far_raw_evaluations = evaluations;
+    }
     if(compareVector(y,optimal)) {
-      optimalFound = true;
+      rawOptimalFound = true;
     }
+    if(instance_id > 1) {
+      transformation.transform_obj_scale(y,instance_id);
+      transformation.transform_obj_shift(y,instance_id);
+    }
+    if(compareObjectives(y,best_so_far_transformed_objectives)){
+      copyVector(y,best_so_far_transformed_objectives);
+      best_so_far_transformed_evaluations = evaluations;
+    }
+    if(compareVector(y,transformed_optimal)) {
+      transformedOptimalFound = true;
+    }
+
+    if(&this->csv_logger != NULL) {
+      (this->csv_logger).write_line(this->evaluations,
```

Comment:
In my opinion, it would be better to call a function of the generic observer here, for example an 'observe' function. This function can then be implemented by the subclasses (csv_logger), which then handles the writing to file / storing to memory / saving to database / ...